src/util/pg.py

```diff
@@ -1621,3 +1622,100 @@ def create_id_sequence(cr, table, set_as_default=True):
                 table=table_sql,
             )
         )
+
+
+def update_table_from_dict(cr, table, columns, mapping, key_col="id", bucket_size=DEFAULT_BUCKET_SIZE):
+    """
+    Update table based on mapping.
+
+    Each `mapping` entry defines the new values for the specified `columns` for the row(s) whose `key_col` value matches the key.
+
+    .. important::
+
+       `columns` can be either a string or a list of strings. Crucially the values of the provided `mapping` must have the same
+       **type** and, if a list, the same **dimensionality** and **order**. Otherwise:
+
+       .. code-block:: python
+
+          columns = ["int_col", "text_col"]
+          mapping = {
+              1: [111, "foo", True],    # third value is ignored
+              2: [222],                 # text_col is set to NULL
+              3: ["bar", 333],          # will attempt to set `int_col` to "bar" and `text_col` to 789
+              4: 444,                   # text_col is set to NULL
+          }
+
+    .. example::
+
+       .. code-block:: python
+
+          # single column update
+          util.update_table_from_dict(
+              cr,
+              "account_move",
+              "always_tax_eligible",
+              {
+                  1: True,
+                  2: False,
+              },
+          )
+
+          # multi-column update
+          util.update_table_from_dict(
+              cr,
+              "account_move",
+              ["closing_return_id", "always_tax_eligible"],
+              {
+                  1: [2, True],
+                  2: [3, False],
+              },
+          )
+
+    .. warning::
+
+       As a side effect, the cursor may be committed.
+
+    :param str table: database's table to perform the update of
+    :param str | list[str] columns: table's columns to update
+    :param dict[any, any | list[any]] mapping: matches values of `key_col` to the new `columns` values
+    :param str key_col: column to match against keys of `mapping`
+    :param int bucket_size: maximum number of rows to update per single query
+    """
+    _validate_table(table)
```

Comment:
I would also like to have a validation of the `mapping` 
```
```suggestion
    _validate_table(table)
    assert isinstance(mapping, dict) and all(isinstance(key, int) and isinstance(value, (list, tuple)) for key, value in mapping.items())
```

Reply:
I think `key_col` should always be an `int`, but I like to have it variable. It would allow to update a table based in a FK instead of a PK. To decide you can grep the repo and see if we currently use the dict trick for other columns than `id`... I think we did but my memory could betray me.

Reply:
From a relatively quick search I found 4 usages of the json trick on columns other than `id`:

* varchar column (2 usages): https://github.com/odoo/upgrade/blob/dad3dd7a1554c526d37fd5f61e6f4a9e18667e56/migrations/sale_subscription/saas~15.3.1.1/pre-migrate.py#L588-L610 and https://github.com/odoo/upgrade/blob/dad3dd7a1554c526d37fd5f61e6f4a9e18667e56/migrations/sale_subscription/saas~15.3.1.1/pre-migrate.py#L612-L634
* varchar column: https://github.com/odoo/upgrade/blob/dad3dd7a1554c526d37fd5f61e6f4a9e18667e56/migrations/account/saas~17.4.1.2/pre-migrate.py#L51-L63
* integer column: https://github.com/odoo/upgrade/blob/dad3dd7a1554c526d37fd5f61e6f4a9e18667e56/migrations/l10n_in_hr_payroll/saas~17.5.1.0/post-migrate.py#L21-L43

Reply:
OK, then maybe it's worth taking the type from the `key_col` in the table directly.

Reply:
That should not be necessary since `JSONB_EACH` returns the key as a text[^1]. 
Unless we want to validate the type of the keys of `mapping`. In which case I would need to define (or reuse, if already available) a mapping from pg types to python types. Wdyt?

[^1]: https://www.postgresql.org/docs/current/functions-json.html#FUNCTIONS-JSON-PROCESSING-TABLE

Reply:
Well I just noticed you are casting the key_col to text before comparing with the actual json dict key, then it's fine. :) What I had in mind was casting the dict key to the column type. In principle my suggestion was to block/fail if we get an invalid json dict (incompatible with the table), I have no strong opinion against current approach TBH. @KangOl wdyt?